numpy/lib/shape_base.py

```diff
@@ -89,9 +90,20 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     outshape = asarray(arr.shape).take(indlist)
     i.put(indlist, ind)
     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-    #  if res is a number, then we have a smaller output array
+
+    # see gh-8363
+    _strcase = False
+    _dtypes = set()
+
+    # If res is a number, then we have a smaller output array.
     if isscalar(res):
```

Comment:
@shoyer , @eric-wieser : I'm failing to see how calling `np.asarray` will help our case here:
~~~python
>>> np.asarray('foo')
array('foo', 
      dtype='<U3')
~~~
Where is the handling of the `object` dtype?  Remember that we also have to handle the case when `res` is multi-dimensional.